\abstract{
Genomic structural variations are an important class of genetic variants with a wide variety of functional impacts. The detection of structural variations using high-throughput short-read sequencing data is a difficult problem, with a lack of algorithm implementations that are able to provide the sensitivity and specificity required in research and clinical settings. Meanwhile, the increasing use of high-throughput sequencing is rapidly generating many large data sets, necessitating the development of algorithms that can provide results rapidly and scale to use available cloud and cluster infrastructures. MapReduce and Hadoop are becoming a standard for managing the distributed processing of large data sets, but existing structural variation detection approaches are difficult to translate into the MapReduce framework. We have developed a general framework for structural variation detection in MapReduce, and demonstrated its potential with Cloudbreak, an implementation that detects genomic deletions and insertions with greater accuracy and much faster runtimes than widely-used existing methods. Cloudbreak's formulation of the problem in terms of feature generation makes it amenable to integrative statistical techniques, allowing the simultaneous integration of many sources of information. We demonstrate the utility of this through an application of conditional random fields, a statistical technique that enables learning conditional probability distributions over labels on sequences of observations. Application of this technique improves Cloudbreak's results, in particular helping its breakpoint resolution. 

In addition to the development of Cloudbreak and its extensions, we describe a data analysis project in which we examined the genomic features that occur near evolutionary breakpoints in the genome of the gibbon, whose karyotype is heavily rearranged compared to other primate species. Using a distributed pipeline to conduct Monte Carlo permutation tests, we find a statistical enrichment of segmental duplications, certain families of transposable elements, and evolutionarily shared binding sites of the protein CTCF near the locations of gibbon rearrangements. These findings may help us understand the process by which SVs formed and were preserved in the gibbon lineage.
}
