\chapter{Analysis of Evolutionary Breakpoint Features}
\label{chap_breakpoint_analysis}

This chapter genomic describes data analysis work unrelated to the algorithmic development of Cloudbreak and its extensions, although it is related at the biological level because it deals with genome breakpoints and rearrangements. It is important not only to be able to locate genomic breakpoints, but also to understand the genomic contexts in which they are located. As we discussed in Chapter~\ref{chap_background}, the sequence features of the genome that neighbor structural variations can leave clues about the mechanisms that caused the rearrangements, as in the cases of large homologies that point to the activity of non-allelic homologous recombination (NAHR) or the microhomologies that indicate non-homologous end joining (NEHJ). In this chapter we examine a set of structural variations that have reshaped a large genome at the evolutionary timescale: the chromosomal rearrangements of the gibbon genome. We analysed the locations in which the gibbon genome has been rearranged relative to humans and the other apes, first using a smaller set of breakpoint data, and then using the entire gibbon genome reference sequence created by the International Consortium for Sequencing and Annotation of the Gibbon Genome. The associations we discovered could help to provide insights into the mechanisms into which the gibbon genome was rearranged. \todo{and perhaps its structure and function}

\section{Background}

In Chapter~\ref{chap_background}, we mentioned that SVs can occur in the genomes of populations of cells at many different levels, ranging from the rearrangements that occur in populations of cells within a tumor in a single individual, to SVs that are polymorphic within the population of a species, and all the way to the structural differences between the genomes of different species. This last type of rearrangement represents SVs that occured within a population and then became fixed in that species and preserved as the species evolved and differentiated. 

Rather than identifying them by comparing short-read sequencing data to a reference genome, as has been the focus of the previous chapters, we can detect these evolutionary breakpoints by directly comparing the genomes of species to each other. In the past this was done by mapping the location of individual genes to particular chromosomes; with the advent of whole genome sequences they be found by conducting sequence alignments (or multiple sequence alignments) of the the reference genomes for the species under study. When species share regions that contain the same genes and other genomic features, in the same order, it is referred to as \emph{shared synteny}. By examining the endpoints of syntenic blocks of the genomes of two species, or \emph{synteny breakpoints}, it is possible to find the breakpoints of the structural variations that rearranged the structure of the two genomes.

Identifying evolutionary breakpoints can give insights into the mechanisms of their formation. It was long thought that chromosomal breakpoints were largely random, such that they are equally likely to occur at any location in the genome~\cite{Ono:1973tr,Nadeau:1984tm}. Higher-resolution analysis, however, revealed that certain regions of the genome are much more likely to give rise to evolutionary breakpoints~\cite{Pevzner:2003ba}, and this result was re-confirmed once whole-genome reference sequences for many species became available~\cite{Murphy:2005hv}. Tantalizingly, Murphy et al. also showed that these reused evolutionary breakpoint regions overlapped significantly with the breakpoints of SVs found in cancer genomes~\cite{Murphy:2005hv}.

This raises the question of whether these regions of the genome have special properties that make them more likely to be observed participating in evolutionary breakpoints. In Chapter~\ref{chap_background}, we described the fact that certain mechanisms for SV formation, in particular NAHR, are a result of having regions with high sequence similarity at different locations in the genome, as the cell's machinery attempts to repair double-stranded breaks in DNA by joining homologous regions together. As one might expect if evolutionary breakpoints were the result of NAHR, a large number of chromosomal rearrangements in mammals, approaching 40\%, are associated with segmental duplications\cite{Bailey:2004fb,Bailey:2006fn}. Closer examination of breakpoints from a variety of species, including human, chimp, macaque, rat, mouse, pig, cattle, dog, opossum, and chicken, confirmed that they contain more sequence that is not unique relative to the rest of the genome than would be expected by chance, including segmental duplications (SDs) and other low-copy number repeats~\cite{Larkin:2009ij}. However, analysis of evolutionary sequence divergence in these repeats indicates that some of the duplications may have occured after the rearrangements, casting dobt on the theory that an NAHR-like process is primarily responsible~\cite{Bailey:2004fb}, and indeed the precise mechanisms behind evolutionary breakpoints remain unknown.

Gibbons represent a unique opportunity to study the process of evolutionary genomic rearrangement because of their highly rearranged genomes. In most branches of the evolutionary tree, large genomic rearrangements are rare events, occurring at a rate of approximately two every ten million years~\cite{Wienberg:2004gt}. However, some species have appear to have faster or slower rates of rearrangments. For example, the orang-utan genome has very few medium and large scale rearrangements relative to the expected number given its time of evolutionary divergence~\cite{Locke:2011gn}. At the other end of the spectrum, gibbon genomes have 10 to 20 times more rearrangements than would be expected given the rate in other mammals, and in fact the four genera of gibbons have numbers of chromosomes ranging from 28 to 52, suggesting that gibbon genome has been shuffled rapidly since it diverged from the other apes 17 to 18 million years ago~\cite{Misceo:2008kg}.

Given their highly rearranged karyotypes, gibbons represent an excellent model in which to study mechanisms of chromosomal rearrangements. Until recently, however, the whole genome sequence of the gibbon was not available. Therefore, several studies analyzed gibbon genome breakpoints using techniques involving bacterial artificial chromosomes (BACs)~~\cite{Girirajan:2009kw,Carbone:2006jk,Carbone:2009p1012}. BACs are medium sized (150-350kb) fragments of DNA from a sample that are isolated, turned into plasmid bacterial chromosomes, and then amplified in a culture. By probing gibbon BACs with arrays or testing their hybridization to human chromosomes using flourescent \emph{in situ} hybridization experiments, it is possible to isolate BACs which span human-gibbon evolutionary breakpoints. This allows one to locate the breakpoint regions to within several hundred kb, and the small size of the BACs relative to the entire genome makes them much easier to sequence, allowing the analysis of sequence features near the breakpoints. Initial studies examined selected BACs from the northern white-cheeked gibbon species \emph{Nomascus leucogenys leucogenys} (NLE) and found a high level of segmental duplications and repetitive sequence at the breakpoints~\cite{Carbone:2006jk,Roberto:2007dt}. Similar results were found in BACs from the white-handed gibbon species \emph{Hylobates lar} (HLE)~\cite{Misceo:2008kg}. A later study increased the scope by examining 24 sequenced NLE BACs that spanned breakpoints, and analyzed gibbon-specific insertions of repeats and segmental duplications at the breakpoints to postulate several possible rearrangement mechanisms~\cite{Girirajan:2009kw}. More recently, Carbone et al.~\cite{Carbone:2009p1012} analyzed 57 NLE BACs and, in addition to segemental duplications, found enrichment for \emph{Alu} repetititve elements. The latter study also found differences in CpG methylation marks on \emph{Alu} elements near the breakpoints, suggesting that there may be an epigenetic process involved in gibbon genomic rearrangements.

\section{Evolutionary Breakpoints in the Gibbon Genome determined by BACs}

To determine whether or not the characteristics observed in the breakpoints identified by BACs from NLE and HLE gibbons were generalizable to the entire gibbon family, we conducted an expanded analysis of gibbon breakpoints based on the sequencing of additional BACs covering breakpoints in the gibbon genera \emph{Symphalangus} and \emph{Hoolock} using samples from the species \emph{Symphalangus syndactylus} (SSY) and \emph{Hoolock leuconedys}. Since each of the gibbon genera has a different karyotype, identifying breakpoints across all four allows for greater insight into the mechanisms and timing of the evolution of the gibbon family. I contributed to this research by anlayzing the overlap between breakpoints from all four gibbon genera with genomic features. This work was published by Capozzi et al.~\cite{Capozzi:2012bb}.

\subsubsection{Methods}

To search for genomic features potentially associated with gibbon chromosomal breakpoints we computed the significance of the overlap between gibbon evolutionary breakpoint regions and a set of genomic features of interest using permutation tests. The purpose of the permutation analysis is to discover the null distribution for the number of overlaps a set of intervals has with a particular feature in the genome. In essence, the test asks: if my intervals were placed in random locations on the genome, what is the probability of seeing the number of overlaps we observed with that features in the actual data? 

To conduct the test, while maintaining the chromosomal assignment and length of break- point regions, we permuted their start coordinates 10,000 times using BEDTools version 2.16.2~\cite{Quinlan:2010km}. Genomic regions annotated as centromeres and telomeres in the ``Gaps'' track of the hg19 build were excluded from possible random placements of the regions. Locations of the features were held constant. We then compared the number of features that overlapped a breakpoint region to the observed distribution of results among the randomly permuted regions, and used the quantile of the real observed value in that distribution as an estimate of the P-value of observing a value equal to or greater than the real observation. In order to be able to conduct a large number of permutations to determine a true background distribution for each feature, we created a pipeline for distributing the permutations across a compute cluster using the grid management system HTCondor.

The analysis was performed on the human hg19 assembly. The features examined were genes, human segmental duplications, and some repeat families (Alu, LINE, ERV, and SVA). We also investigated the associations between breakpoint regions and chromatin structure by testing the overlap with open chromatin regions in human embryonic stem cells reported by the ENCODE consortium~\cite{ENCODEProjectConsortium:2011iz}. 

\subsubsection{Results}

We found a significant enrichment for genes (Bonferroni adjusted P-value = 0.0287), human segmental duplications (P = 0.0366), Alu (P < 0.0001), and SVA (P = 0.0008) (Figure~\ref{gibbon_bac_permutations}). We did not find significant enrichment for LINE and ERV repeats, nor for the ENCODE open chromatin regions. Systematically shifting the location of breakpoint regions by increments of 10 kb up- and downstream of their actual location, up to a maximum of 1 MB, shows that the locations of the breakpoint regions gives the greatest or close to the greatest number of overlaps with the four significantly overlapping features (genes, segmental duplications, Alu, and SVA) in the local genomic neighborhood, also shown in Figure~\ref{gibbon_bac_permutations}.

\begin{figure}
\centering
\includegraphics{figures/gibbon_bac_permutations.jpg}
\caption{Enrichment of genomic features in breakpoint regions. Permutation tests were used to assess the overlap between the gibbon breakpoints and genomic features. (A) Segmental duplications; (B) Alu elements; and (C) SVA elements. The black vertical line indicates the observed value for the breakpoints identified in the study. In all three cases it is evident that the genomic features have a higher overlap with the breakpoints than one could expect by chance.}
\label{gibbon_bac_permutations}

\end{figure}

\section{Analysis of Breakpoints from the Gibbon Genome Reference Sequence}

Continuing the work I have already done in analyzing the evolutionary breakpoints mapped with the gibbon BAC clones, I am working with the International Consortium for Sequencing and Annotation of the gibbon genome to analyze the additional breakpoints discovered through the assembly of the gibbon reference genome. To do so, I am conducting additional association tests with a variety of features. One of these features is a set of binding sites of the evolutionarily conserved binding factor CTCF, which I have identified through the analysis of chromatin immunoprecipitation followed by sequencing (ChIP-seq) data. CTCF has recently been shown to be associated with the three-dimensional structure of DNA and chromatin in the cell~\cite{Dixon:2012gc}, and therefore may have important associations with genomic DNA breakpoints and structural variations.

\todo{I am also helping to analyze the breakpoints from a set of over forty mesenchymal tumor samples in which genomic rearrangements have created ring-shaped chromosomes. The breakpoints that cause these chromosomal rearrangements cluster into several frequently broken areas. I am currently working on analyzing these breakpoint regions using my permutation analysis pipeline in order to determine genomic features that may be significantly associated with these breakpoints, including repeat elements, open chromatin sites and other data from ENCODE, and binding motifs for enzymes known to be involved with breakpoint formation.}

\section{Breakpoints in the gibbon genome reference}

\subsection{Overlap with genomic features: repeats and genes}

To analyze the enrichment of genomic features in the regions flanking evolutionary breakpoints, we used a permutation based approach. The number of overlaps between breakpoint flanks and each feature of interest in the assembled genome (the observed overlap count) was compared to a background distribution calculated by randomly permuting the locations of breakpoint regions 100,000 times. The Nleu1.1 version of the assembly was used for these analyses. For our breakpoint regions, we chose those breakpoints for which we had single nucleotide resolution, and added flanking regions to either side of the breakpoint; in breakpoints in which the breakpoint fell within a gibbon-specific repeat element, we chose the flanking regions of the repeat. In each permutation, the location of each breakpoint region was randomly changed, while keeping the length and scaffold assignment of the breakpoint region the same. We then counted the number of overlaps between the randomized breakpoint regions and the feature of interest (the permuted overlap count). Enrichment p-values were computed as the proportion of permuted overlap counts that were more extreme than the observed overlap count. We also visualized the spatial relationship of breakpoint regions to each type of figure by simultaneously shifting the locations of the breakpoint regions up to 1MB in each direction, in increments of 25kb, and counting the proportion of shifted breakpoint regions that overlapped a feature of interest, after discarding regions that were shifted beyond the beginning or end of a scaffold. Permutation testing and shift testing were carried out using custom Python scripts and the BEDtools~\cite{Quinlan:2010km}, pybedtools\cite{Dale:2011cl}, and BEDOPS~\cite{Neph:2012kq} libraries; the code is publicly available at \url{https://github.com/cwhelan/permuting-feature-enrichment-test}.

We tested for enrichment in the breakpoint regions of the following features: genes, segmental duplications, and several classes of repetitive elements: Alu, L1, LAVA, and LTR. In addition to testing the entire Alu family, we also tested the subfamilies AluS, AluJ, and AluY individually. Gene locations were taken from Ensembl build 70. For the segmental duplication analysis, we used the segmental duplications identified by the WSSD method (Section S3). Alu, L1, and LTR locations were identified using RepeatMasker output. In order to determine the distance from the breakpoints at which enrichments are strongest, we varied the size of the breakpoint flanking regions by adding differently sized intervals; we tested flanking regions of size 100bp, 250bp, 500bp, and 1000bp. We corrected for multiple testing using the FDR under dependency method of Bejamini and Yekutieli~\cite{Benjamini:2001fs}. Breakpoint regions are depleted for genes, but enriched for Alu elements and segmental duplications (Extended data figure 2). The enrichment for Alu is primarily due to a strong enrichment of the AluS subfamily.Even though genes were depleted overall, some of the genes overlapping with breakpoints belong to interesting biological categories (Table ST5.1 ).

In addition to testing the count of overlaps between breakpoint flanking regions and repeats, we also conducted a complementary test that examined the distance of each breakpoint to the nearest repeat of a given class. For this test we used only the 44 breakpoints for which we had single nucleotide breakpoint resolution, with no Gibbon-specific repeats directly intersecting the breakpoint. We compared these breakpoint locations to 10,000 randomly selected regions in the Nleu1.0 genome, selected using the randomBed program from the BEDTools suite. The distance to the nearest repeat for either the breakpoints or random positions was determined using BEDTools closestBed.  We compared the distribution of distances to a repeat for the breakpoints to the distribution of distances to a repeat for the 10,000 randomly selected positions using the Kolmogorov-Smirnov (K-S) test. We examined the distance to any repeat, as well as those for Alu, LINE, and LTR elements, and finally the AluJ, AluS, and AluY subfamilies. After FDR correction for multiple hypothesis testing, the distance test showed similar results to the overlap test described above, with signficant results for the Alu family as a whole, and for the AluJ and AluS subfamilies, indicating that the breakpoints tend to be closer to those repeats than random locations in the genome.

Breakpoint regions were simultaneously shifted in increments of 25kb, up to a maximum of 1MB in each direction, and the proportion of breakpoint regions that overlap a feature of interest is reported. Shifts show that breakpoints are centered on regions that are depleted in genes but close to regions that contain genes, while the opposite is true for segmental duplications. Alu elements are more evenly spread across the shift regions (Extended data figure 2).

\subsection{Overlap with CTCF binding sites}

\subsubsection{Chromatin Immunoprecipitation (ChIP) sequencing for CTCF}

CTCF ChIP-seq assays were performed according to Schmidt et al.~\cite{Schmidt:2012dt} on eight EBV-transformed lymphoblastoid cell lines established for the same individuals used for the diversity panel (Table ST2.1). In brief, CTCF-bound DNA was immunoprecipitated using an Anti-CTCF rabbit polyclonal antibody (07-729, Millipore). End-repair was performed on immunoprecipitated and input DNA prior to A-tailing and ligation to single-end Illumina sequencing adapters. DNA was amplified using Illumina primers 1.1 and 2.1 in an 18-cycle PCR reaction. Gel electrophoresis was used to select 200-300 bp DNA fragments. DNA libraries were sequenced using 36 bp reads on an Illumina Genome Analyser II according to the manufacturer’s instructions.

\subsubsection{Peak calling from CTCF ChIP-seq Data}

We aligned reads to the nomLeu2 reference using BWA (version 0.62)~\cite{Li:2009p836} with default parameters, and removed non-uniquely mapping reads. We then called peaks using CCAT~\cite{Xu:2010fu}, with parameters fragmentSize 100, slidingWindowSize 150, movingStep 10, isStrandSensitiveMode 1, minCount 10, minScore 4.0, and bootstrapPass 50. We then combined the peaks called across the different individuals and chose the following set for further analysis: any peak called in an individual individual by CCAT with an FDR of less than 0.05, as well as any peak that was called in more than one individual with an FDR of less than 0.1. (Table ST5.2)

\subsubsection{Determination of Gibbon-specific and shared CTCF binding sites}

Gibbon peaks were classified as shared or gibbon-specific by comparing them to a set of CTCF peaks called on human, macaque, and orangutan individuals (Schwalie, P.C. et al. under review). First, orthologous locations of gibbon CTCF peaks in other primate species were determined using a local installation of the Ensembl Compara multi-species alignment database (http://uswest.ensembl.org/info/docs/api/compara/index.html). This database contains alignments of the references for human (GRCh37), chimp (CHIMP2.1.4), gorilla (gorGor3.1), orangutan (PPYG2), rhesus macaque (MMUL\_1), and gibbon (Nleu1.1). A multi-species alignment of each Gibbon CTCF peak region was generated in paml format. Gibbon CTCF peaks for which no multi-alignment were present, or where the nucleotide alignment identity to human and macaque was less than 70\%, were excluded from analysis. We then created a non-redundant list of non-gibbon CTCF peaks by converting the human, orangutan and macaque peaks to gibbon coordinates using the Compara database, and then merging the results using the mergeBed tool from the BEDTools suite to remove redundant entries. Shared and gibbon-specific CTCF peaks were then identified as those gibbon CTCF peaks that did or did not intersect a peak in the non-redundant list of non-gibbon CTCF peaks.

\subsubsection{Analysis of CTCF peaks in relationship with gibbon-human synteny breakpoints}

We identified 52,685 CTCF binding sites. Because of CTCF's function as an insulator and its known association with the boundaries of DNA topological domains~\cite{Dixon:2012gc}, we tested the overlap of CTCF binding sites and the gibbon synteny breakpoint regions. We find that 12 of the 1kb regions flanking breakpoints overlap with CTCF binding sites (one example showed in Extended data figure 2). Using the same permutation analysis described above, this overlap has an enrichment p-value of 0.0009. This effect was even stronger when we expanded the breakpoint flanking regions to 25kb,  as 95 expanded flanking regions overlap CTCF peaks for an enrichment p-value <0.0001 (Extended data figure 2).

We then tested whether the CTCF peaks causing the enrichment in breakpoint regions are specific to gibbons. Using CTCF ChIP-seq data for samples from human, macaque, and orangutan individuals, we classified CTCF binding sites as unique to gibbons (11,449 sites) or shared with a primate ancestor (41,236 sites). We found that the gibbon breakpoint regions are heavily enriched for CTCF binding sites shared with a primate ancestor (enrichment p-value = 0.0006) but are not significantly enriched for gibbon-specific CTCF binding sites. Again, the enrichment is stronger in the 25kb expanded breakpoint regions (enrichment p-value <0.0001) for shared binding sites, but not for gibbon-specific binding sites. This suggests that the formation or selection of gibbon genome rearrangements was associated with ancestral CTCF binding sites.